### PR TITLE
WIP: Diagnostics: Symbol Access

### PR DESCRIPTION
**Pull request by nordlow**
## Infrastructure for symbol referencing/access and partial dataflow analysis to catch

- unused variables
- read but unmodified variables that should be declared `const` (or immutable)
- dereferencing of (potentially) null pointers and classes

False positives for symbols referenced in templates will be elided.

The goal of making this a PR at this stage in development is to prevent me from wasting time on analysis that has already been implemented in dmd but I'm unaware of.

As much logic as possible will be moved to functions in `diagnostics.d` to make it less intrusive to the existing files.

For a preview of what features already work see the added files `test/fail_compilation/diag_un*.d`.

**Spec**:
- Each `Dsymbol` is classified as being referenced (by name) or not (unused).
- Each `VarDeclaration` is classified as being (also) either read, written, or both (`VRef`).
- Each such verdict uses a tri-state boolean `Trilean` being either `no`, `yes` or `maybe`.
- Analsis will happen in functions in `diagnostics.d` that are called at suitable `visit()` functions either in `expressionsem.d` or `statementsem.d`.
- Reporting happens in functions in `diagnostics.d` that are currently called in its own visitor. To reduce overhead this visitor should removed and calls should be placed in existing or new `visit()`-overloads in `semantic3.d`.

**Notes**:
- Detects instance function-local imports that are not chosen in favour of global imports of same name in UFCS-calls.

**Open Question**:
- How should hints on how to elide the warning messages be formatted? Currently this happens on the same line making lines a bit long. Is it better to use an alternative formatting to provide the fix-it hint on a new line using `warningSupplemental` instead?

**Tasks**:
- Currently: standard imports such as `import std.algorithm.searching` will give false negatives if several level modules import the same module but this is a good start which will make the experience confusing. See `ss.isReferenced = true;` in `ScopeDsymbol.search()` in `dmodule.d`.
- Should we do any further special handing of unused `virtual` (`override`) member function parameters?
- Unused template parameters: More relevant to warn for than run-time parameters.
- Shall we exclude diagnostics for all members of all `extern` structs/classes?
- Module-level private symbols may be referenced solely in templates and/or unittests, versions, etc. If so, their unusedness can only be verified when all the templates in the same module is expanded and unittests are enabled. One way to detect this is to scan the parse tree (or is it lex-stream) for identifiers that equal any unused symbols and tag those that match as used. Further, AFAICT, if a string import or mixin occur inside a template we cannot be certain that any at all symbol is unused. Furthermore, unittests 
- Can we detect when a whole package is being sent in one batch to dmd? If so we could enable some _innovative_ detection of
   - library-wide symbols qualified as `package` and
   - application-wide symbols qualified as `package` or `public`?!
- Many unittests have unused local functions and variables that receive the return value of code that is verified to compile. What to do with those?
- Should we have a special naming convention for eliding diagnostics, for instance `_`? For instance foreach over range expressions require a variable: `foreach (i; 0 .. 10)`.
- When `foreach`-statements are lowered to functions, special hidden variable names are emitted, for instance `__key$N` and `__limit$N` and `__r$N`, `__aggr$N`. I've excluded these by simple filtering out those variables whose `ident.toChars()` start with `__`.
- Overridden virtual class member functions needs special care. As they can be called via superclass-pointers and those must be included in the isReferenced-check and only if non-of them are called can we be certain that virtual member is not called.
- Detect unused labels by checking how a `GotoStatement`s resolve label `Identifier`.
- Check only if `-unittest` is given or no unittests are present in module
- Detect unused `package` declarations when building whole packages (via dmd flag passed by `dub`)
- Detect unmodified mutable `package` declarations when building whole packages (via dmd flag passed by `dub`)
- Diagnose: unmodified declaration (member) should be qualified as `static immutable` or `enum`
- Diagnose: unmodified normal variables with value semantics should be qualified as `const` or `immutable`
- Diagnose: unmodified normal variables with reference semantics should be qualified as `immutable` or `const`
- Diagnose: advice to prefer `immutable` before `const` for reference variables that are used either in defining scope only or passed as `const scope`.
- Correct `loc` of parameters for `FuncDeclaration` to point to the parameter name `Identifier` instead of `FuncDeclaration.loc`.

Other code-changes:
- Makes some members of `SemanticTimeTransitiveVisitor` `final`.

### CODE REVIEW

File: src/build.d [hunk 1]
```diff
@@ -1176,7 +1176,7 @@ auto sourceFiles()
             parse.d parsetimevisitor.d permissivevisitor.d printast.d safe.d sapply.d scanelf.d scanmach.d
             scanmscoff.d scanomf.d semantic2.d semantic3.d sideeffect.d statement.d statement_rewrite_walker.d
             statementsem.d staticassert.d staticcond.d stmtstate.d target.d templateparamsem.d traits.d
-            transitivevisitor.d typesem.d typinf.d utils.d visitor.d vsoptions.d foreachvar.d
+            transitivevisitor.d typesem.d typinf.d utils.d visitor.d vsoptions.d foreachvar.d unused.d
```
Comment (andralex) [inline src/build.d:1179]: This is a confusing name because people will think the module is actually not used. 

```suggestion
            transitivevisitor.d typesem.d typinf.d utils.d visitor.d vsoptions.d foreachvar.d detect_unused.d
```

Comment (nordlow) [reply src/build.d]: Fixed.

File: src/build.d [hunk 2]
```diff
@@ -1176,7 +1176,7 @@ auto sourceFiles()
             parse.d parsetimevisitor.d permissivevisitor.d printast.d safe.d sapply.d scanelf.d scanmach.d
             scanmscoff.d scanomf.d semantic2.d semantic3.d sideeffect.d statement.d statement_rewrite_walker.d
             statementsem.d staticassert.d staticcond.d stmtstate.d target.d templateparamsem.d traits.d
-            transitivevisitor.d typesem.d typinf.d utils.d visitor.d vsoptions.d foreachvar.d
+            transitivevisitor.d typesem.d typinf.d utils.d visitor.d vsoptions.d foreachvar.d detect_unused.d
```
[the remark this answers is not in the document]
Comment (nordlow) [reply src/build.d:1179]: What about putting this in a subdir 'diag'?

File: src/dmd/detect_unused.d [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    case Prot.Kind.public_:
+        return "public";
+    case Prot.Kind.export_:
+        return "export";
+    }
+}
+
+static private bool shouldWarnUnunsed(in Prot.Kind protKind) @safe pure nothrow @nogc
+{
+    final switch (protKind)
```
Comment (UplinkCoder) [inline src/dmd/detect_unused.d:258]: You can group the cases.

**Comment by nordlow** [comment]
I'm making some interesting progress now with reasonable low percentage of false positives. I suggest hiding this behind a new flag for now, say `-w=ununsed` or `-diagnose=unused`. What do you think, @UplinkCoder?

**Comment by UplinkCoder** [comment]
Yes a flag. 
Perhaps -preview=warn-unused.

File: src/dmd/detect_unused.d (hunk 3 above)
Comment (nordlow) [reply src/dmd/detect_unused.d:258]: Ahh, thanks. This could also be detect in a diagnostics! ;)

**Comment by UplinkCoder** [comment]
@nordlow 
Looks like your group cases, returns false only.

**Comment by nordlow** [comment]
> @nordlow
> Looks like your group cases, returns false only.

Thanks.

**Comment by nordlow** [comment]
Can you take a look at the questions, @UplinkCoder?

**Comment by wilzbach** [comment]
> Yes a flag. 
> Perhaps -preview=warn-unused.

It's bike shedding, but as you plan to add more diagnosis features, I would recommend to go for a full flag like `-diagnostics` as the user can then do `-diagnostics=all`.

File: src/build.d (hunk 2 above)
[the remark this answers is not in the document]
Comment (wilzbach) [reply src/build.d:1179]: `Diag` is a bit cryptic, I would prefer `diagnostics`.

**Comment by nordlow** [comment]
> > Yes a flag.
> > Perhaps -preview=warn-unused.
> 
> It's bike shedding, but as you plan to add more diagnosis features, I would recommend to go for a full flag like `-diagnostics` as the user can then do `-diagnostics=all`.

I'm leaning more towards allowing `-w=unused' which will mean cleaner branching on warning and diagnostics state.

**Comment by nordlow** [comment]
This is starting become useful now.

It detected:

```
/home/per/Work/dmd/src/dmd/dsymbol.d(1146,13): Warning: unused match expression `ti` of if statement
```

```
in 
    {
        Dsymbol s = parent;
        for (; s; s = s.toParent())
        {
            if (auto ti = s.isTemplateInstance())
            {
                return false;
            }
            if (auto m = s.isModule())
            {
                if (!m.isRoot())
                    return true;
                break;
            }
        }
        return false;
    }
```

File: src/build.d (hunk 2 above)
[the remark this answers is not in the document]
Comment (nordlow) [reply src/build.d]: Done.

**Comment by nordlow** [comment]
This is what things looks like in Emacs+FlyCheck with this branch analysing one the tests.

Screenshot from 2020-08-27 11-51-06

**Comment by nordlow** [comment]
Seems like the line diagnostics isn't consistent over different platforms. See for instance the diff:

```
 fail_compilation/diag_unused.d(109): Warning: unused member (enumerator) `b` of private enum `PrivateEnum2`
 fail_compilation/diag_unused.d(110): Warning: unused member (enumerator) `c` of private enum `PrivateEnum2`
 fail_compilation/diag_unused.d(115): Warning: unused private alias `PrivateUnusedInt`
-fail_compilation/diag_unused.d(122): Warning: unused private alias `PrivateUInt`
+fail_compilation/diag_unused.d(127): Warning: unused private alias `PrivateUInt`
 fail_compilation/diag_unused.d(132): Warning: unused private variable `px` of module
 fail_compilation/diag_unused.d(137): Warning: unused local variable `x` of function, remove, rename it to `_` or prepend `_` to its name to silence
 fail_compilation/diag_unused.d(138): Warning: unused local variable `y` of function, remove, rename it to `_` or prepend `_` to its name to silence
```

**Comment by nordlow** [comment]
> Could you explain a bit what this is supposed to do, and when the warning is supposed to trigger? @nordlow

Look at the test files in the commit for a set of warning cases that are currently handled.

**Comment by wilzbach** [comment]
@nordlow I strongly recommend getting the current state of things merged instead and then opening an individual PR for each diagnostic task as otherwise this PR will keep grow and be very hard to review and maintain.

**Comment by nordlow** [comment]
I'm planning to split things up. The meaning of this PR is to enable people to try this diagnostics out themselves so they can make comments on its usefulness with emphasis on false negatives and positives.

**Comment by wilzbach** [comment]
Well you could already filter out some basics like for example the PR to add the diagnostics flag.

> Do you know how I can avoid putting load on the CI

Well you need to put this in the commit, not title.
It's possible for some CIs e.g. https://buildkite.com/docs/pipelines/ignoring-a-commit, but I don't think all CI handle this special tag (or for some it needs to be configured).

I wouldn't worry so much about it. We don't have a big CI load these days and you probably want to ensure that your work compiles and passes on all platforms anyhow (and if compilation fails the CI will quickly fail and abort).

**Comment by nordlow** [comment]
Ok, lets start with the flag.

Ok to hide all this behind `-preview=diagnostics`?
Then I think it would be nice to be able to select individual diagnostics aswell. In the long run I think

- `w=unused`
- `w=unused-import`
- `w=all`
...

is the cleanest.

Should we add something like that aswell now and allow it only when `-preview=diagnostics` is set?

File: src/dmd/dimport.d [hunk 4]
```diff
@@ -11,6 +11,8 @@
 
 module dmd.dimport;
 
+import core.stdc.stdio;
```
Comment (RazvanN7) [inline src/dmd/dimport.d:14]: unrelated change

File: src/dmd/dimport.d [hunk 5]
```diff
@@ -145,8 +147,7 @@ extern (C++) final class Import : Dsymbol
                 return true;
             }
         }
-        Dsymbol s = dst.lookup(id);
-        if (s)
+        if (Dsymbol s = dst.lookup(id))
```
Comment (RazvanN7) [inline src/dmd/dimport.d:150]: please do not mix refactorings with feature implementation

File: src/dmd/declaration.d [hunk 6]
```diff
@@ -1087,6 +1089,24 @@ extern (C++) final class OverDeclaration : Declaration
     }
 }
 
+/** Tri-state bool.
+ *
+ * See_Also: https://en.wikipedia.org/wiki/Three-valued_logic
+ */
+enum trool : ubyte
```
Comment (RazvanN7) [inline src/dmd/declaration.d:1096]: I would sugest using a more descriptive name. Also, the dmd norm seems to be that enum declarations start with a capital letter.

**Review by RazvanN7 — COMMENTED** [review]
I haven't thoroughly reviewed this so my comments are actually nitpicks, however, I would have expected that the implementation is located in the search function of ScopeDsymbol : all symbols are searched for their declarations and once they are found an `isUsed` field should be updated. However, from what I could tell the implementation is scattered over multiple places. Why is that?

File: src/dmd/statementsem.d [hunk 7]
```diff
@@ -201,6 +202,19 @@ private extern (C++) final class StatementSemanticVisitor : Visitor
             s.exp = checkGC(sc, s.exp);
             if (s.exp.op == TOK.error)
                 return setError();
+
+            if (doCheckUnused())
```
Comment (RazvanN7) [inline src/dmd/statementsem.d:206]: So you are marking the variable as referenced only if it was assigned? What about imported variables that have been passed as arguments to a function?

File: src/dmd/statementsem.d [hunk 8]
```diff
@@ -201,6 +202,19 @@ private extern (C++) final class StatementSemanticVisitor : Visitor
...
+            {
+                if (AssignExp ae = s.exp.isAssignExp)
```
Comment (RazvanN7) [inline src/dmd/statementsem.d:208]: This code can go inside the expression visitor in `visit(AssignExp)`

File: src/dmd/statementsem.d (hunk 7 above)
Comment (nordlow) [reply src/dmd/statementsem.d:206]: > Could you explain a bit what this is supposed to do, and when the warning is supposed to trigger? @nordlow

See the tests to see what cases are currently covered.

File: src/dmd/dimport.d (hunk 4 above)
Comment (wilzbach) [reply src/dmd/dimport.d:14]: This is fine and commonly accepted to be added in PRs.

File: src/dmd/dimport.d (hunk 5 above)
Comment (nordlow) [reply src/dmd/dimport.d:150]: Done.

File: src/dmd/dimport.d (hunk 4 above)
Comment (nordlow) [reply src/dmd/dimport.d:14]: Done.

File: src/dmd/statementsem.d (hunk 7 above)
Comment (nordlow) [reply src/dmd/statementsem.d:206]: I'm checking three kinds of referencing; by name, by value read, by value written.

**Comment by nordlow** [comment]
Hmm. I don't understand why the C++ build fails. I can't find any mismatches for the C++ bindings.

File: src/dmd/statementsem.d (hunk 8 above)
Comment (nordlow) [reply src/dmd/statementsem.d:208]: visit(AssignExp exp) is long with many returns so I'm highly unsure where in I should inject my code.

File: src/dmd/declaration.d (hunk 6 above)
Comment (nordlow) [reply src/dmd/declaration.d]: Done.

**Comment by WalterBright** [comment]
This appears to be a large number of only loosely related changes lumped together. It would be far more reviewable to do incrementally.

**Comment by WalterBright** [comment]
> I don't understand why the C++ build fails.

This is why doing small, incremental PRs is much more feasible - it makes it much easier to track down the causes of these sorts of problems.

**Comment by nordlow** [comment]
> > I don't understand why the C++ build fails.

I had mismatches in the .h files.

**Comment by nordlow** [comment]
> This appears to be a large number of only loosely related changes lumped together. It would be far more reviewable to do incrementally.

I'm building an infrastructure for symbol referencing an partial dataflow analysis. Many things are interlinked. And I'm far from done. Before I have the whole picture I won't know how to split it up into a set of simple pull requests. The goal of putting this in a PR is to prevent me from wasting time on analysis that has already been implemented in dmd. I'm gonna move as much logic as possible to functions in `diagnostics.d` to make it less intrusive to the existing files.

**Comment by dkorpel** [comment]
Like Walter said:

> This appears to be a large number of only loosely related changes lumped together. It would be far more reviewable to do incrementally.

I don't see a 5000 line PR like this getting in, but perhaps future PRs can reboot parts of this code
